//------------------------------------------------------------------------------
// protobuf specification for network security objects
//------------------------------------------------------------------------------

syntax = "proto3";

import "types.proto";
import "qos.proto";

package nwsec;

service NwSecurity {
    rpc SecurityProfileCreate (SecurityProfileRequestMsg) returns (SecurityProfileResponseMsg) {}
    rpc SecurityProfileUpdate (SecurityProfileRequestMsg) returns (SecurityProfileResponseMsg) {}
    rpc SecurityProfileDelete (SecurityProfileDeleteRequestMsg) returns (SecurityProfileDeleteResponseMsg) {}
    rpc SecurityProfileGet (SecurityProfileGetRequestMsg) returns (SecurityProfileGetResponseMsg) {}

    rpc SecurityGroupCreate (SecurityGroupMsg) returns (SecurityGroupResponseMsg) {}
    rpc SecurityGroupUpdate (SecurityGroupMsg) returns (SecurityGroupResponseMsg) {}

    rpc SecurityPolicyRuleCreate (SecurityPolicyRuleMsg) returns (SecurityPolicyRuleResponseMsg) {}
    rpc SecurityPolicyRuleUpdate (SecurityPolicyRuleMsg) returns (SecurityPolicyRuleResponseMsg) {}
}

// normalization feature actions upon detecting malformed packets
enum NormalizationAction {
  NORM_ACTION_NONE    = 0;
  NORM_ACTION_ALLOW   = 1;    // allow the packet as is
  NORM_ACTION_DROP    = 2;    // drop the packet
  NORM_ACTION_EDIT    = 3;    // reset/trim/fix invalid fields/flags
}

// SecurityProfileKeyHandle is used to operate on a security profile
// either by its id or HAL allocated handle
message SecurityProfileKeyHandle {
  oneof key_or_handle {
    uint32     profile_id     = 1;     // App allocated unique security profile id
    fixed64    profile_handle = 2;     // id of the SecurityProfile returned by HAL
  }
}

// SecurityProfileSpec object captures feature knobs that are most likely
// shareable across L2 segments, tenants, or even interfaces and most of the
// time operates on defaults. It is possible that there is only object of this
// kind in the system.
// NOTE: user is not expected to fill in all the fields explicitly, that job is
//       done by the agent. Agent will be exposing user visible knobs and fill
//       in the defaults when interacting with HAL.
message SecurityProfileSpec {
  types.ObjectMeta             meta                            = 1;    // common object meta
  SecurityProfileKeyHandle     key_or_handle                   = 2;
  bool                         cnxn_tracking_en                = 3;
  bool                         ipsg_en                         = 4;
  bool                         tcp_rtt_estimate_en             = 5;
  uint32                       session_idle_timeout            = 6;
  uint32                       tcp_cnxn_setup_timeout          = 7;
  uint32                       tcp_close_timeout               = 8;
  uint32                       tcp_close_wait_timeout          = 9;

  bool                         ip_normalization_en             = 10;
  bool                         tcp_normalization_en            = 11;
  bool                         icmp_normalization_en           = 12;

  // IP normalization knobs
  bool                         ip_ttl_change_detect_en         = 20;
  NormalizationAction          ip_rsvd_flags_action            = 21;
  NormalizationAction          ip_df_action                    = 22;
  NormalizationAction          ip_options_action               = 23;
  NormalizationAction          ip_invalid_len_action           = 24;
  uint32                       ip_normalize_ttl                = 25;

  // ICMP/ICMPv6 normalization knobs
  NormalizationAction          icmp_invalid_code_action        = 30;
  bool                         icmp_deprecated_msgs_drop       = 31;
  bool                         icmp_redirect_msg_drop          = 32;

  // TCP normalization knobs
  bool                         tcp_non_syn_first_pkt_drop      = 40;
  bool                         tcp_syncookie_en                = 41;
  bool                         tcp_split_handshake_detect_en   = 42;
  bool                         tcp_split_handshake_drop        = 43;
  NormalizationAction          tcp_rsvd_flags_action           = 44;
  NormalizationAction          tcp_unexpected_mss_action       = 45;
  NormalizationAction          tcp_unexpected_win_scale_action = 46;
  NormalizationAction          tcp_urg_ptr_not_set_action      = 47;
  NormalizationAction          tcp_urg_flag_not_set_action     = 48;
  NormalizationAction          tcp_urg_payload_missing_action  = 49;
  NormalizationAction          tcp_rst_with_data_action        = 50;
  NormalizationAction          tcp_data_len_gt_mss_action      = 51;
  NormalizationAction          tcp_data_len_gt_win_size_action = 52;
  NormalizationAction          tcp_unexpected_ts_option_action = 53;
  NormalizationAction          tcp_unexpected_echo_ts_action   = 54;
  bool                         tcp_ts_not_present_drop         = 55;
  bool                         tcp_invalid_flags_drop          = 56;
  bool                         tcp_flags_nonsyn_noack_drop     = 57;
}

// SecurityProfileRequestMsg is batched add or modify profile request
message SecurityProfileRequestMsg {
  repeated SecurityProfileSpec    request = 1;    // batched request
}

// SecurityProfile operational status
message SecurityProfileStatus {
  fixed64    profile_handle = 1;    // id of the security profile returned by HAL
}

// SecurityProfileResponse is response to SecurityProfileSpec
message SecurityProfileResponse {
  types.ApiStatus          api_status     = 1;    // API status code
  SecurityProfileStatus    profile_status = 2;    // SecurityProfile profile operational status
}

// SecurityProfileResponseMsg is batched response to SecurityProfileRequestMsg
message SecurityProfileResponseMsg {
  repeated SecurityProfileResponse    response = 1;    // batched response
}

// SecurityProfileDeleteRequest is used to delete a SecurityProfile profile
message SecurityProfileDeleteRequest {
  types.ObjectMeta            meta          = 1;    // object meta
  SecurityProfileKeyHandle    key_or_handle = 2;    // SecurityProfile profile being deleted
}

// SecurityProfileDeleteRequestMsg is used to delete a batch of SecurityProfile profiles
message SecurityProfileDeleteRequestMsg {
  repeated SecurityProfileDeleteRequest request = 1;    // batched delete request
}

// SecurityProfileDeleteResponseMsg is batched response to SecurityProfileDeleteRequestMsg
message SecurityProfileDeleteResponseMsg {
  repeated types.ApiStatus    api_status = 1;    // API status code
}

// SecurityProfileGetRequest is used to get information about a L2 Segment
message SecurityProfileGetRequest {
  types.ObjectMeta            meta          = 1;    // object meta
  SecurityProfileKeyHandle    key_or_handle = 2;    // SecurityProfile profile key or handle
}

// SecurityProfileGetRequestMsg is batched GET request for SecurityProfile profiles
message SecurityProfileGetRequestMsg {
  repeated SecurityProfileGetRequest    request = 1;    // batched get request
}

// SecurityProfileStats is the statistics object for each SecurityProfile profile
message SecurityProfileStats {
}

// SecurityProfileGetResponse captures all the information about a SecurityProfile profile
message SecurityProfileGetResponse {
  SecurityProfileSpec      spec   = 1;
  SecurityProfileStatus    status = 2;
  SecurityProfileStats     stats  = 3;
}

// SecurityProfileGetResponseMsg is the batched response to SecurityProfileGetRequestMsg
message SecurityProfileGetResponseMsg {
  repeated SecurityProfileGetResponse    response = 1;    // batched get response
}

// SecurityGroupSpec captures the configuration parameters for a security group
message SecurityGroupSpec {
  types.ObjectMeta    meta     = 1;    // common object meta
  fixed32             sg_id    = 2;    // security group id
}

// SecurityGroupMsg is batched add or modify security group request
message SecurityGroupMsg {
  repeated SecurityGroupSpec    request = 1;    // batched request
}

// security group operational status
message SecurityGroupStatus {
  fixed64    sg_handle = 1;    // id of the security group returned by HAL
}

// SecurityGroupResponse is the response to SecurityGroupSpec
message SecurityGroupResponse {
  types.ApiStatus        api_status  = 1;    // API status code
  SecurityGroupStatus    status      = 2;    // operational status
}

// SecurityGroupResponseMsg is batched response to SecurityGroupMsg
message SecurityGroupResponseMsg {
  repeated SecurityGroupResponse    response = 1;    // batched response
}

// Service represents an application service defined in terms of protocol and port
message Service {
   types.IPProtocol    ip_protocol = 1;    // IP protocol
   uint32              port        = 2;    // TCP or UDP port, 0 for other protocols
}

// security policy actions
enum Action {
 SECURITY_POLICY_ACTION_NONE   = 0;
 SECURITY_POLICY_ACTION_ALLOW  = 1;
 SECURITY_POLICY_ACTION_DENY   = 2;
 SECURITY_POLICY_ACTION_REJECT = 3;
}

// SecurityPolicyRule defines a rule between a pair of security groups
// A rule can contain multiple services to match against
message SecurityPolicyRuleSpec {
  types.ObjectMeta meta        = 1;    // object meta
  fixed32          src_sg_id   = 2;    // source security group id
  fixed32          dst_sg_id   = 3;    // destination security group id
  repeated Service svc         = 4;    // application service
  Action           action      = 5;    // action to take
  bool             rule_log_en = 6;    // log if this rule is hit
  qos.QOSActions   qos_actions = 7;    // QOS actions 
}

// SecurityPolicyRuleMsg is batched add or modify security policy request
message SecurityPolicyRuleMsg {
  repeated SecurityPolicyRuleSpec    request = 1;    // batched request
}

// SecurityPolicyRuleStatus is the operational state of a rule
message SecurityPolicyRuleStatus {
  fixed64    rule_handle = 1;    // id of policy rule returned by HAL
}

// SecurityPolicyRuleResponse is the response to SecurityPolicyRuleSpec
message SecurityPolicyRuleResponse {
  types.ApiStatus             api_status  = 1;    // API status code
  SecurityPolicyRuleStatus    status      = 2;    // operational status
}

// SecurityPolicyRuleResponseMsg is batched response to SecurityPolicyRuleMsg
message SecurityPolicyRuleResponseMsg {
  repeated SecurityPolicyRuleResponse    response = 1;    // batched response
}
